full-node/db/sov-schema-db/src/snapshot.rs

```diff
@@ -0,0 +1,158 @@
+//! Snapshot related logic
+
+use std::sync::{Arc, LockResult, Mutex, RwLock, RwLockReadGuard};
+
+use crate::schema::{KeyCodec, ValueCodec};
+use crate::{Operation, Schema, SchemaBatch, DB};
+
+/// Id of database snapshot
+pub type SnapshotId = u64;
+
+/// A trait to make nested calls to several [`Schema`]
+pub trait QueryManager {
+    /// Get a value from snapshot or its parents
+    fn get<S: Schema>(
+        &self,
+        snapshot_id: SnapshotId,
+        key: &impl KeyCodec<S>,
+    ) -> anyhow::Result<Option<Operation>>;
```

Comment:
I wonder if `Option<Operation>` is a leaky abstraction in this case. Do we really need to distinguish between an absent value and a deleted value?

Reply:
Yes, the distinction is important, because if value is absent, then snapshot needs to query database, but if it was deleted, "None" is the correct response, even if value is still in database.

There might be different approach tho, when snapshot only reads from local and parent, but manager have to read from the database if value wasn't found in snapshots.

Reply:
Ah, yes! I totally missed that.

Reply:
Pros for `QueryManager` to read from database:

 * There's single instance of QueryManager, which means only single handle to the database.
 * It is ReadLock, so many readers should not block each other
 
 Cons:
 
  * When QueryManager will be flushing snapshot to rocksdb, it will block all readers. It will block them in current design, but for less time. 
  
I wonder if benchmark is needed.